ow_plexil/scripts/identify_sample_location.py

```diff
@@ -0,0 +1,243 @@
+    try:
+      image = self.bridge.cv2_to_imgmsg(self.sample_location_image, "bgr8")
+    except CvBridgeError as err:
+      rospy.logerror("CV Bridge error: {0}".format(err))
+      return
+    self.publish_photo.publish(image)
+
+
+  def identify_sample_spots(self, location_image, filter_type):
+    '''Uses OpenCv to process the given image. For filter type Brown we isolate a range of brown colors 
+    on the image and draw the contours. For Dark filter (the default filter) we isolate the darkest spots 
+    in the image (usually shadows) and find the countours. The 2d center location is taken for these contours
+    and are sorted in descending order, we then draw/store the largest 5 contours/center points to return.'''
+    try:
+      image = self.bridge.imgmsg_to_cv2(location_image, "bgr8")
+    except CvBridgeError as err:
+      rospy.logerror("CV Bridge error: {0}".format(err))
+      return None, None, None
+
+    #defaults to Dark Spot if argument doesnt match or none is given
+    if filter_type == "Brown":
+     #filtering brown spots on image
+      hsv = cv.cvtColor(image, cv.COLOR_BGR2HSV)
+      lower = np.array([0,16,32])
+      upper = np.array([19,255,255])
+      thresh = cv.inRange(hsv, lower, upper)
+    else:
+      #filtering dark spots on image
+      gray = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
+      ret, thresh = cv.threshold(gray, 15, 255,cv.THRESH_BINARY_INV)
+
+    #find contours and sort them largest to smallest
+    _, contours, hierarchy = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
+    contours.sort(reverse=True, key=cv.contourArea)
+    sample_points_2d = []
+    contour_areas = []
+
+    #draw all contours over a certain size; 5000 seems good for now, only taking the largest 5
+    for i in contours:
+      area = cv.contourArea(i)
+      if area >= 5000 and area <= 1000000:
+        moment = cv.moments(i)
+        cx = int(moment["m10"] / moment["m00"])
+        cy = int(moment["m01"] / moment["m00"])
+        cv.drawContours(image, [i], 0, 255, 3)
+        sample_points_2d.append((cx, cy))
+        contour_areas.append(area)
+      if len(contour_areas) >= 5:
+        break
+
+    return sample_points_2d, contour_areas, image
+
+
+  def get_3d_sample_point(self, point_cloud, sample_points_2d, contour_areas):
+    '''Given a list of 2d points we find the corresponding 3d point within the pointcloud. 
+    Takes the first valid point (largest contour with valid point) and transforms it into the
+    base_link frame. Returns the new 3d sample point, its corresponding contour and the original
+    2d location on the image.'''
+    if sample_points_2d:
+      site_coordinate = None
+      index = 0
+      #finds corresponding 3d point in point cloud from 2d pixel point takes the largest (first valid) point
+      for i in sensor_msgs.point_cloud2.read_points(point_cloud, uvs=sample_points_2d):
+        if(np.isnan(i[0]) or np.isnan(i[1]) or np.isnan([2])):
```

Comment:
Outer parentheses are not needed, also note the third is isnan seems incorrect
```suggestion
        if np.isnan(i[0]) or np.isnan(i[1]) or np.isnan(i[2]):
```

Reply:
Good catch, thanks!